extension/agenthealth/handler/stats/provider/statuscode.go

```diff
@@ -0,0 +1,171 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package provider
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/amazon-contributing/opentelemetry-collector-contrib/extension/awsmiddleware"
+
+	"github.com/aws/amazon-cloudwatch-agent/extension/agenthealth/handler/stats/agent"
+)
+
+const (
+	statusResetInterval = 5 * time.Minute
+	statusHandlerID     = "cloudwatchagent.StatusCodeHandler"
+)
+
+var (
+	statusCodeSingleton *StatusCodeHandler
+	statusCodeStatsOnce sync.Once
+)
+
+// StatusCodeHandler provides monitoring for status codes per operation.
+type StatusCodeHandler struct {
+	statsByOperation sync.Map
+	resetTimer       *time.Timer
+	filter           agent.OperationsFilter
+}
+
+// GetStatusCodeStats retrieves or initializes the singleton StatusCodeHandler.
+func GetStatusCodeStats(filter agent.OperationsFilter) *StatusCodeHandler {
+	statusCodeStatsOnce.Do(func() {
+		handler := &StatusCodeHandler{}
+		handler.filter = filter
+		handler.startResetTimer()
+		statusCodeSingleton = handler
+	})
+	return statusCodeSingleton
+}
+
+// startResetTimer initializes a reset timer to clear stats every 5 minutes.
+func (h *StatusCodeHandler) startResetTimer() {
+	ticker := time.NewTicker(statusResetInterval)
+
+	go func() {
+		for range ticker.C {
+			h.statsByOperation.Clear()
+			log.Println("Status code stats reset.")
+		}
+	}()
+}
+
+// HandleRequest is a no-op for the StatusCodeHandler.
+func (h *StatusCodeHandler) HandleRequest(ctx context.Context, _ *http.Request) {}
+
+// HandleResponse processes the HTTP response to update status code stats.
+func (h *StatusCodeHandler) HandleResponse(ctx context.Context, r *http.Response) {
+	// Extract the operation name
+	operation := awsmiddleware.GetOperationName(ctx)
+	if !h.filter.IsAllowed(operation) {
+		log.Printf("Operation %s is not allowed", operation)
+		return
+	} else {
+		log.Printf("Processing response for operation: %s", operation)
+	}
+
+	operation = GetShortOperationName(operation)
+	if operation == "" {
+		return
+	}
+	statusCode := r.StatusCode
+
+	value, loaded := h.statsByOperation.LoadOrStore(operation, &[5]int{})
+	if !loaded {
+		log.Printf("Initializing stats for operation: %s", operation)
+	}
+	stats := value.(*[5]int)
+
+	h.updateStatusCodeCount(stats, statusCode, operation)
+
+	h.statsByOperation.Store(operation, stats)
```

Comment:
Hmm, these lines consist of a read-modify-write operation. Two threads running this at the same time could clobber each others updates. Only the latter thread will have it's updates applied. As an example:

1. Thread A and B are both handling an HTTP response for the same operation. HTTP response is 200 for both threads.
1. Thread A loads current values, let's say stats[0] = 10
1. Thread B loads current values, also sees stats[0] = 10
1. Both threads now have a copy stats array
1. B adds 1 to its copy, stats[0] = 11
1. B stores its result
1. A adds 1 to it's copy, stats[0] = 11
1. A stores its result

We end up with stats[0]=11 for the operation instead of stats[0]=12. Perhaps its not actually an issue though. Is it possible we could be handling more than one HTTP response for the same operation at once? If not, then I wouldn't worry about it. If we don't know, or if it doesn't happen now but could in the future, we should handle it.